src/pagination.rs

```diff
@@ -0,0 +1,312 @@
+            self.next_index = index + 1;
+            self.next_index_permalink = Some(interpret_permalink(&config, &doc, index + 1)?);
+        }
+        Ok(())
+    }
+}
+
+impl Into<liquid::value::Object> for Paginator {
+    fn into(self) -> liquid::value::Object {
+        let mut object = liquid::value::Object::new();
+        // if no pages, means we have indexes instead, `tags` like cases for exemple
+        if let Some(pages) = self.pages {
+            object.insert("pages".into(), liquid::value::Value::Array(pages));
+        }
+        // list of available indexes, in `tags` like cases
+        if let Some(indexes) = self.indexes {
+            object.insert(
+                "indexes".into(),
+                liquid::value::Value::Array(
+                    indexes.iter().map(liquid::value::Value::scalar).collect(),
+                ),
+            );
+        }
+        object.insert(
+            "index".into(),
+            liquid::value::Value::scalar(self.index as i32),
+        );
+        object.insert(
+            "index_permalink".into(),
+            liquid::value::Value::scalar(self.index_permalink),
+        );
+        if let Some(previous_index_permalink) = self.previous_index_permalink {
+            object.insert(
+                "previous_index".into(),
+                liquid::value::Value::scalar(self.previous_index as i32),
+            );
+            object.insert(
+                "previous_index_permalink".into(),
+                liquid::value::Value::scalar(previous_index_permalink),
+            );
+        }
+        if let Some(next_index_permalink) = self.next_index_permalink {
+            object.insert(
+                "next_index".into(),
+                liquid::value::Value::scalar(self.next_index as i32),
+            );
+            object.insert(
+                "next_index_permalink".into(),
+                liquid::value::Value::scalar(next_index_permalink),
+            );
+        }
+        object.insert(
+            "first_index_permalink".into(),
+            liquid::value::Value::scalar(self.first_index_permalink),
+        );
+        object.insert(
+            "last_index_permalink".into(),
+            liquid::value::Value::scalar(self.last_index_permalink),
+        );
+        object.insert(
+            "total_indexes".into(),
+            liquid::value::Value::scalar(self.total_indexes as i32),
+        );
+        object.insert(
+            "total_pages".into(),
+            liquid::value::Value::scalar(self.total_pages as i32),
+        );
+        object
+    }
+}
+
+pub fn generate_paginators(
+    doc: &mut Document,
+    posts_data: &[liquid::value::Value],
+) -> Result<Vec<Paginator>> {
+    let config = doc.front.pagination.as_ref().unwrap();
+    let mut all_posts = posts_data.to_vec();
+    match config.include {
+        Include::All => {
+            sort_posts(&mut all_posts, &config);
+            create_all_paginators(&mut all_posts, &doc, &config)
```

Comment:
So the reason you use `&mut` seems to be
- We divide `all_posts` into chunks
- We move those chunks out rather than clone them,  for performance

My concerns
- mutable data makes the code harder to reason about (non-local reasoning)
- Prevents things like rayon which I want to eventually sprinkle in

One idea I have is to make `all_posts` a `Vec<&Value>`.  That might make sorting faster (just moving around a reference but at the cost of an extra indirection) and then the `clone` becomes explicit in `set_current_index_info` rather than implicit in here.
```rust
let all_posts: Vec<_> = posts_data.iter().collect();
```

When I wrap up my work with CLI-WG and Liquid refactorings, I want to come back to Cobalt and switch its architecture to an ECS.  It'd solve all the awkward architecture issues I've been having and allow us some nice reductions in cloning like we do here.

Reply:
If use this
```rust
let all_posts: Vec<_> = posts_data.iter().collect();
```
`sort_posts` will need to return a new `Vec`, am I right?

Reply:
I'd say not to worry too hard about it right now.

Reply:
Oops, I've done something, let me know if it's not too ugly, otherwise I'll revert